### ISSUE TITLE
Rewrite to be based on php-build

### ISSUE THREAD

**Issue by smorimoto** [asdf-community/asdf-php#72]
There is no benefit on our side to managing this too complex shell nightmare, and it should be managed centrally if possible.

**Comment by moljac024** [comment]
Better yet, why not install php precompiled? That would make the plugin so much easier to use (no need to install build dependencies and wait for php to compile etc.)

Side rant: I can't believe it's 2021 and we still haven't shaken this "compile everything" idea in general oss. Right now, right this very second, there's millions of compiles of the same thing happening all over the world.

**Comment by smorimoto** [comment]
I don't mind. PR is always welcome.

**Comment by moljac024** [comment]
Perhaps supporting both initially is the way to go. How about we transition to php-build but we cache the builds of individual versions somewhere online? So when someone wants to install a specific version, we:

- Check if we have the build for this architecture cached already and if we do, just pull that and be done
- If we don't have it cached, do the build via php-build and upload the artifacts to the build cache for the next person to benefit

Then the obvious question is, where can we host the builds for free?

What do you think?

**Comment by smorimoto** [comment]
The biggest issue is how to distribute the cache as you mentioned. Also, I'm not really familiar with PHP, but are they really relocatable? At this point, I don't have a clear ETA for php-build migration, so if you want to work on it, I recommend that you start by migrating to it first.

**Comment by moljac024** [comment]
Sadly I'm not familiar with PHP either. My only contribution was the comment about avoiding compilation if possible but in retrospect it's not really useful.

**Comment by smorimoto** [comment]
In fact, there is no incentive to host the cache at all, and implementing something to make sure the cache is safe is quite a hassle. And we need to minimize the dependencies that it requires, and I don't want to do that. I won't stop if someone does it in another plugin, but I absolutely don't want to distribute third-party caches to users through this plugin.

**Comment by iloveitaly** [comment]
I looked into this a bit. IMHO this codebase is way better organized, and actually supports a greater set of options than php-build. I'd vote for *not* moving over to use php-build.

**Comment by smorimoto** [comment]
The advantage of using php-build is that it can support more versions. Our current code only targets the latest version and does not support older versions. (There are some versions that can be built, but we can't say that they are **officially** supported.)

**Comment by iloveitaly** [comment]
True, although there are some tradeoffs to consider. The downside is there are a lot of compilation options which are not supported in php-build (`--with-gmp` for instance). For me, this would be really problematic. Additionally, php-build doesn't seem to fix into the `asdf` super cleanly (the default assumption is you are using `phpenv`, although there could be an easy way around that. 

We *could* expand the PHP version matrix in this github workflow to cover more PHP versions, although we still wouldn't have the larger test suite available to us and there would still be duplicated efforts.

**Comment by smorimoto** [comment]
> The downside is there are a lot of compilation options which are not supported in php-build (--with-gmp for instance).
I don't think so: https://github.com/php-build/php-build/blob/master/man/php-build.1.ronn#environment

**Comment by b1rdex** [comment]
Another option: https://github.com/shivammathur/homebrew-php + https://github.com/shivammathur/homebrew-extensions

**Comment by smorimoto** [comment]
Somehow I had to come back to the PHP world. I've read some PR, but I'm sure what we need now is to rewrite the php-build based plugin. Does anyone want to tackle it?

**Comment by Tarik02** [comment]
> Better yet, why not install php precompiled? That would make the plugin so much easier to use (no need to install build dependencies and wait for php to compile etc.)
> 
> Side rant: I can't believe it's 2021 and we still haven't shaken this "compile everything" idea in general oss. Right now, right this very second, there's millions of compiles of the same thing happening all over the world.

If you want to provide precompiled PHP, you have to also provide precompiled versions of many libraries used by PHP, for example OpenSSL, icu and others since PHP depends on them and user's system can have variety of versions (especially major versions).

**Comment by smorimoto** [comment]
There are currently not enough main maintainers for this plugin, so it would be nice if someone could do it. Furthermore, if anyone would like to replace this with a php-build-based implementation, I'd be happy to help.

**Comment by Tarik02** [comment]
> There are currently not enough main maintainers for this plugin, so it would be nice if someone could do it. Furthermore, if anyone would like to replace this with a php-build-based implementation, I'd be happy to help.

I did some investigation, I still have CI broken for example (https://github.com/Tarik02/asdf-php). However this works for my local machine (Arch) it does not work on latest ubuntu machines. As I discovered, the problem is that ubuntu includes libxml which itself uses libicu (which I built together with php) of different version.

So, now I see the only approach is to build most of dependencies before building PHP and use them, e.g. whole prefix with dependencies for it. I'm gonna do more investigation about this.

For arch, it builds (assuming you have the needed dependencies) most verions like 5.6.40, 7.0.33, 7.1.33, 7.4, 8.1.10 and they work as expected. But for ubuntu it's completely broken.

**Comment by a21ns1g4ts** [comment]
there are a lot of deprecated packages required by php <= 7. No easy way to work with them.

**Comment by roylindauer** [comment]
Is php-build actively maintained? There are issues with it now that prevent any version of php being installed and the maintainers have been silent on issues about it. As a result I currently cannot install php at all using mise/asdf on macos because of that provider issue.

I do not know what the path forward is or how to use this plugin in its current state.

**Comment by Rusydy** [comment]
Hi all, 

I almost done with my PR; to use php-build for compiling.
After a few more test on my linux and mac os I will merge it.

If you all want to review or comment on my PR I do appreciate it 

Thank you

**Comment by Rusydy** [comment]
**2026 January 24th, Update** 

Since 2026 January 1st I cannot install PHP 8.0.0 on macOS, then turns out the error is in php-build, I need to work on php-build first
